pkgs/stdenv/booter.nix

```
@@ -98,20 +98,26 @@ stageFuns: let
     thisStage =
       if args.__raw or false
       then args'
-      else allPackages ((builtins.removeAttrs args' ["selfBuild"]) // {
-        adjacentPackages = if args.selfBuild or true then null else rec {
-          pkgsBuildBuild = prevStage.buildPackages;
-          pkgsBuildHost = prevStage;
-          pkgsBuildTarget =
-            if args.stdenv.targetPlatform == args.stdenv.hostPlatform
-            then pkgsBuildHost
-            else assert args.stdenv.hostPlatform == args.stdenv.buildPlatform; thisStage;
-          pkgsHostHost =
-            if args.stdenv.hostPlatform == args.stdenv.targetPlatform
-            then thisStage
-            else assert args.stdenv.buildPlatform == args.stdenv.hostPlatform; pkgsBuildHost;
-          pkgsTargetTarget = nextStage;
-        };
+      else allPackages (args' // {
+        # Technically we only need to check `build!=target` since nixpkgs currently doesn't allow
+        # both `build!=host` and `host!=target` at the same time, but we check both in order to be
+        # future-proof.
+        actuallySplice = with thisStage.stdenv; buildPlatform != hostPlatform || hostPlatform  != targetPlatform;
+        adjacentPackages =
+          let
+            # search backwards from `stage` for the most recent stage containing packages which execute on `on` and target `for`
+            seek = stage: { on, for }@args:
+              if on  == stage.stdenv.hostPlatform &&
+                 for == stage.stdenv.targetPlatform
+              then stage
+              else seek stage.stdenv.__bootPackages args;
```

So the boot packages relation is like a linked list. What happens if it isn't found? Surely that would be a Nixpkgs internal error, but perhaps reporting it would help out the next person to understand what kind of mistake was made.

```suggestion
              else seek stage.stdenv.__bootPackages or (throw "nixpkgs internal error: Bootstrapping stage with requested host and target platform not found.") args;
```

I don't know if a lack of `__bootPackages` signifies the end of the list, so maybe the above doesn't make sense.

If the end is not observable we'd be looking at a stack overflow error. If you have an alternate solution that is more easily shown to terminate, that might be nice, but I'd accept this anyway because I think it would only happen to Nixpkgs developer who work on the bootstrapping code, and I have a draft nix PR to add stack traces to stack overflows (unlike infinite recursions which already have them). So the risk isn't that bad.

---

> I don't know if a lack of `__bootPackages` signifies the end of the list, so maybe the above doesn't make sense.

Technically the earliest stage is the `bootstrap-files`, and it lacks a `__bootPackages` attribute (it is the only stage which lacks this attribute) -- the check you've added is useful here.